### PR TITLE
Create `Build retention` page. Update `Build export` page

### PR DESCRIPTION
**Pull request by mitchbne**
# Description

## Build Retention

We want to keep a page about **Build retention** (and this should be separate to **Build exports).** 
This page should talk about retention periods and perhaps a lifecycle around when data is exported (i.e. maybe it happens at the time of a build being archived, but that's thinking very futuristically). 

This page should discuss **Build exports** as a concept/feature available for Enterprise plans and link to the https://buildkite.com/docs/pipelines/build-exports page where Enterprise customers can read more about how to provide us with an AWS S3 bucket and how to configure it.

image

## Build Exports
Build exports page added to Buildkite docs which includes:

- [x] What the build exports feature is
- [x] Flow diagram showing how build exports works
- [x] Who can use it
- [x] How a customers build data is stored
- [x] How to configure your own S3 bucket, will guidelines on how to do that (as well as useful links)
- [x] Enabling build exports via the Buildkite UI

## Context

Figma file with the diagrams for build exports and build retention
https://www.figma.com/file/hkC2zVsM9VQ3LfGO9jcHyH/Workflow-diagrams-for-build-export%2Fretention?node-id=1-160&t=VIrXn1lFnlYjx8Ny-0

- Follow up to #2024.
- [PDP-720](https://linear.app/buildkite/issue/PDP-720/documentation)

### CODE REVIEW

**Comment by dabarrell** [comment]
I reckon it's worth detailing the retention periods for Open Source and Non-Profit plans here too :) Both for users of those plans, and because it looks good...

File: pages/pipelines/build_export.md [hunk 1]
```diff
@@ -1,12 +1,35 @@
...
+    annotations.json.gz
+    artifacts.json.gz
+    build.json.gz
+    step-uploads.json.gz
+    jobs/
+        job-{UUID}.json.gz
+        job-{UUID}.log
+```
+
+The files will be stored in the following formats: [Annotations](https://buildkite.com/docs/apis/rest-api/annotations#list-annotations-for-a-build), [Artifacts](https://buildkite.com/docs/apis/rest-api/artifacts#list-artifacts-for-a-build), [Builds](https://buildkite.com/docs/apis/rest-api/builds#get-a-build), [Jobs](https://buildkite.com/docs/apis/rest-api/jobs#retry-a-job).
```
Comment (pda) [inline pages/pipelines/build_export.md:30]: `/docs/apis/rest-api/jobs#retry-a-job` shouldn't link to the “Retry a job” section…

But it's kinda weird that there's no “Get a build” section in the Jobs REST API docs. I guess that's because there's no API for that! API::JobsController#show doesn't exist… jobs are only shown embedded in Builds… which also seems weird to me, but that's a whole other thing.

Maybe for Jobs we should say “as would be embedded in a [Build via the REST API](https://buildkite.com/docs/apis/rest-api/builds#get-a-build)” rather than linking to the Jobs API.

Oh and maybe for Build we should say “but without `jobs`, which are stored in separate files.”

Comment (123sarahj123) [reply pages/pipelines/build_export.md]: Sweet. added your suggestion :)

File: pages/pipelines/build_exports.md [hunk 2]
```diff
@@ -1,12 +1,39 @@
...
+If you do not configure an S3 bucket, the exported build data will continue to be stored automatically by Buildkite. You can not access this build data using the Buildkite UI or API, but you can request them from us at any time. Buildkite will store this build data for 12 months before deleting.
+
+## How it works
+Builds older than the build retention limit will be automatically exported (in JSON format) to either the S3 bucket you have provided, or Buildkite will continue to store that build data (in JSON format) for a further 12 months before deletion.
+
+<%= image "build-exports-flow-chart.png", alt: "Simplified flow chart of the build exports process" %>
+
+You are able to configure you own S3 bucket in the UI. See [Enabling build exports](#enabling-build-exports) below.
+
+Buildkite stores your build data as multiple JSON files which include the following data:
```
Comment (pda) [inline pages/pipelines/build_exports.md:22]:
```suggestion
Buildkite exports each build as multiple gzipped JSON files which include the following data:
```

**Review by pda — COMMENTED** [review]
Looking good! I added some thoughts.

File: pages/pipelines/build_exports.md [hunk 3]
```diff
@@ -1,12 +1,39 @@
-# Export build data
+# Build exports
 
 Each [Buildkite plan](https://buildkite.com/pricing) has a maximum build retention
-period, after which your oldest builds are automatically deleted.
+period. Once your builds reach the retention period, they will be removed from Buildkite.
```
[the remark this answers is not in the document]
Comment (123sarahj123) [reply pages/pipelines/build_exports.md:4]: done

File: pages/pipelines/build_exports.md [hunk 4]
```diff
@@ -1,12 +1,39 @@
...
 
 > 📘 Enterprise feature
-> Build export is only available on an [Enterprise](https://buildkite.com/pricing) plan.
+> Build exports is only available on an [Enterprise](https://buildkite.com/pricing) plan, which has a build retention period of 12 months.
+
+If you need to retain build data beyond the retention period in your [Buildkite plan](https://buildkite.com/pricing), you can use the build exports feature.
```
[the remark this answers is not in the document]
Comment (123sarahj123) [reply pages/pipelines/build_exports.md:9]: done :D

**Comment by dabarrell** [comment]
One thought - the placement of the diagram currently makes it look like it belongs to the "Exporting build data" heading. Perhaps it makes sense to move it above that heading, so it instead belongs to "Retention periods". Or even above retention periods? not sure

**Comment by dabarrell** [comment]
@123sarahj123 I think I worked out why the diagram looks a bit strange to me, on this page. The Figma "section" (the purple box) is an added layer that isn't really necessary (including the purple "Build Retention" header, and adds visual noise. Could you try just exporting the two diagrams on a transparent background instead? (or white? @mbelton-buildkite might have a preference)

Dropping the purple box also means the diagram itself can be bigger, which would help with the lack of a "zoom in" function

**Comment by 123sarahj123** [comment]
@dabarrell I gave that a go but the Figma export feature seems to add a lot of padding. 
image

I resorted to just cropping the png down:
image

image

**Comment by 123sarahj123** [comment]
Here is the figma file with the diagrams if they require further tinkering:
https://www.figma.com/file/hkC2zVsM9VQ3LfGO9jcHyH/Workflow-diagrams-for-build-export%2Fretention?node-id=1-160&t=VIrXn1lFnlYjx8Ny-0

File: pages/pipelines/build_exports.md [hunk 5]
```diff
@@ -1,12 +1,36 @@
...
-period, after which your oldest builds are automatically deleted.
+# Build exports
 
 > 📘 Enterprise feature
-> Build export is only available on an [Enterprise](https://buildkite.com/pricing) plan.
+> Build exports is only available on an [Enterprise](https://buildkite.com/pricing) plan, which has a [build retention](/docs/pipelines/build-retention) period of 12 months.
+
+If you need to retain build data beyond the [retention period](/docs/pipelines/build-retention) in your [Buildkite plan](https://buildkite.com/pricing), you can use the build exports feature.
+
+Build exports enables you to configure your own S3 bucket for Buildkite to export your build data to.
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:8]: I'm unsure about referring to everything as a named feature, so let's avoid that here. I think we can also combine these two sentences:

```suggestion
If you need to retain build data beyond the [retention period](/docs/pipelines/build-retention) in your [Buildkite plan](https://buildkite.com/pricing), you can export the data to your own [Amazon S3 bucket](https://aws.amazon.com/s3/).
```

Note, I also referred to the full Amazon name and linked to it since this is the first mention.

File: pages/pipelines/build_exports.md [hunk 6]
```diff
@@ -1,12 +1,36 @@
...
+
+If you do not configure an S3 bucket, the exported build data will be stored for 18 months by Buildkite in case you need it. You cannot access this build data using the Buildkite UI or API, but you can request them from us by contacting support.
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:10]: Style:

```suggestion
If you don't configure an S3 bucket, Buildkite stores the build data for 18 months in case you need it. You cannot access this build data through the API or Buildkite dashboard, but you can request the data by contacting support.
```

File: pages/pipelines/build_exports.md [hunk 7]
```diff
@@ -1,12 +1,36 @@
...
+
+## How it works
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:12]: Include a blank line between headings and paragraphs:

```suggestion
## How it works

```

File: pages/pipelines/build_exports.md [hunk 8]
```diff
@@ -1,12 +1,36 @@
...
+Builds older than the build retention limit will be automatically exported as JSON to the S3 bucket you have configured. If you haven't configured a build export bucket, Buildkite will store that build data as JSON in our own S3 bucket for a further 18 months in case you need it.
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:13]: A few small style nitpicks + adding a lead-in sentence for the diagram.

```suggestion
Builds older than the build retention limit are automatically exported as JSON to the S3 bucket you have configured. If you haven't configured a bucket for build export, Buildkite stores that build data as JSON in our own S3 bucket for a further 18 months in case you need it. The following diagram outlines this process.
```

File: pages/pipelines/build_exports.md [hunk 9]
```diff
@@ -1,12 +1,36 @@
...
+
+<%= image "build-exports-flow-chart.png", alt: "Simplified flow chart of the build exports process" %>
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:15]: The diagram says that Buildkite stores data fro 12 months if I don't have a bucket configured, but the text says 18 months. Which is right?

File: pages/pipelines/build_exports.md [hunk 10]
```diff
@@ -1,12 +1,36 @@
...
+
+You are able to configure you own S3 bucket in the UI. See [Enabling build exports](#enabling-build-exports) below.
+
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:18]: I don't think this is necessary since the page it pretty simple.

```suggestion
```

File: pages/pipelines/build_exports.md [hunk 11]
```diff
@@ -1,12 +1,36 @@
...
+Buildkite exports each build as multiple gzipped JSON files which include the following data:
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:19]: Missing comma:

```suggestion
Buildkite exports each build as multiple gzipped JSON files, which include the following data:
```

File: pages/pipelines/build_exports.md [hunk 12]
```diff
@@ -1,12 +1,36 @@
...
+    annotations.json.gz
+    artifacts.json.gz
+    build.json.gz
+    step-uploads.json.gz
+    jobs/
+        job-{UUID}.json.gz
+        job-{UUID}.log
+```
+
+The files will be stored in the following formats: [Annotations](https://buildkite.com/docs/apis/rest-api/annotations#list-annotations-for-a-build), [Artifacts](https://buildkite.com/docs/apis/rest-api/artifacts#list-artifacts-for-a-build), [Builds](https://buildkite.com/docs/apis/rest-api/builds#get-a-build) (but without `jobs` as they are stored in separate files), and Jobs (as would be embedded in a [Build via the REST API](https://buildkite.com/docs/apis/rest-api/builds#get-a-build)).
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:32]: I find this sentence quite hard to parse. How about using a bulleted list instead?

```suggestion
The files are stored in the following formats: 

* [Annotations](https://buildkite.com/docs/apis/rest-api/annotations#list-annotations-for-a-build)
* [Artifacts](https://buildkite.com/docs/apis/rest-api/artifacts#list-artifacts-for-a-build)
* [Builds](https://buildkite.com/docs/apis/rest-api/builds#get-a-build) (but without `jobs`, as they are stored in separate files)
* Jobs (as would be embedded in a [Build via the REST API](https://buildkite.com/docs/apis/rest-api/builds#get-a-build))
```

Comment (mbelton-buildkite) [reply pages/pipelines/build_exports.md:32]: Actually, we've talked about the file format as JSON and gzip already. This part is about the file contents, right? Maybe:

```suggestion
The file contents match the following objects:

* [Annotations](https://buildkite.com/docs/apis/rest-api/annotations#list-annotations-for-a-build)
* [Artifacts](https://buildkite.com/docs/apis/rest-api/artifacts#list-artifacts-for-a-build)
* [Builds](https://buildkite.com/docs/apis/rest-api/builds#get-a-build) (but without `jobs`, as they are stored in separate files)
* Jobs (as would be embedded in a [Build via the REST API](https://buildkite.com/docs/apis/rest-api/builds#get-a-build))
```

File: pages/pipelines/build_exports.md [hunk 13]
```diff
@@ -1,12 +1,36 @@
...
 
-If you need to retain build data beyond the retention period in your [Buildkite plan](https://buildkite.com/pricing), you can have Buildkite export the data to a private Amazon S3 bucket. Prior to old build data being removed, Buildkite exports JSON representations of the builds to the S3 bucket you provide.
 
 ## Preparing your S3 bucket
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:35]: I don't remember this section going in. Maybe I missed a PR 🤔

Should this be part of enabling build exports?

Thinking something like:

```
## Configure build exports

### Prepare your S3 bucket

### Enable build exports
```

**Review by mbelton-buildkite — COMMENTED** [review]
A bunch of suggestions, but nothing too major. I need to look over the diagrams still, but thought I'd publish these comments first.

**Comment by mbelton-buildkite** [comment]
> Here is the figma file with the diagrams if they require further tinkering: https://www.figma.com/file/hkC2zVsM9VQ3LfGO9jcHyH/Workflow-diagrams-for-build-export%2Fretention?node-id=1-160&t=VIrXn1lFnlYjx8Ny-0

Could someone give me write access to the file? I'd like to play around with them a bit.

**Comment by dabarrell** [comment]
>I resorted to creating a new figma file and adding the diagrams as frames.

@123sarahj123 I imagine we'd like want the white background instead of purple :) Would you mind changing that?

File: pages/pipelines/build_exports.md [hunk 14]
```diff
@@ -1,12 +1,36 @@
...
+
+```
+buildkite/build-exports/org={UUID}/date={YYYY-MM-DD}/pipeline={UUID}/build={UUID}/
+    annotations.json.gz
+    artifacts.json.gz
+    build.json.gz
+    step-uploads.json.gz
+    jobs/
+        job-{UUID}.json.gz
+        job-{UUID}.log
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:29]: Thoughts on using some more visual cues for the file structure?

```suggestion
buildkite/build-exports/org={UUID}/date={YYYY-MM-DD}/pipeline={UUID}/build={UUID}/
├── annotations.json.gz
├── artifacts.json.gz
├── build.json.gz
├── step-uploads.json.gz
└── jobs/
    ├── job-{UUID}.json.gz
    └── job-{UUID}.log
```

Made with https://ascii-tree-generator.com/

**Comment by mbelton-buildkite** [comment]
Let me know if you want to pair on this to move faster 😊

**Comment by 123sarahj123** [comment]
@mbelton-buildkite You are a legend ! Thank you so much for the thorough review! 
If you had the time to pair, feel free! Otherwise I am more than happy to crank them out now :D

File: images/docs/pipelines/build_exports/build-exports-flow-chart.png
[the remark this answers is not in the document]
Comment (123sarahj123) [reply images/docs/pipelines/build_exports/build-exports-flow-chart.png:1]: sure thing. I'll get onto this now

File: pages/pipelines/build_exports.md (hunk 9 above)
Comment (123sarahj123) [reply pages/pipelines/build_exports.md:13]: Good catch. I'll clear this up

Comment (123sarahj123) [reply pages/pipelines/build_exports.md]: when the builds are exported (and if they have not provided us a bucket) we store their build data for another 18 months. I corrected the diagram :D

File: pages/pipelines/build_exports.md (hunk 13 above)
Comment (123sarahj123) [reply pages/pipelines/build_exports.md:37]: Have formatted it as you suggested :)

File: pages/pipelines/build_retention.md [hunk 15]
```diff
@@ -0,0 +1,57 @@
+# Build retention
+
+Each [Buildkite plan](https://buildkite.com/pricing) has a maximum build retention period. Once your builds reach the retention period, they will be removed from Buildkite.
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_retention.md:3]: Sorry, I missed commenting on this in the last round. Style nitpicks:

```suggestion
Each [Buildkite plan](https://buildkite.com/pricing) has a maximum build retention period. Once builds reach the retention period, their data is removed from Buildkite.
```

File: pages/pipelines/build_retention.md [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    </tr>
+    <tr>
+      <td>Enterprise plan</td>
+      <td>1 year</td>
+      <td>Yes</td>
+    </tr>
+  </tbody>
+</table>
+
+## Exporting build data
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_retention.md:53]: Add blank line after heading:

```suggestion
## Exporting build data

```

File: pages/pipelines/build_exports.md [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+├── annotations.json.gz
+├── artifacts.json.gz
+├── build.json.gz
+├── step-uploads.json.gz
+└── jobs/
+    ├── job-{UUID}.json.gz
+    └── job-{UUID}.log
+```
+
+The files are stored in the following formats:
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:29]: This isn't rendering correctly because there's a blank line missing here:

```suggestion
The files are stored in the following formats:

```

Screenshot 2023-04-27 at 5 29 30 pm

File: pages/pipelines/build_exports.md [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+* [Annotations](https://buildkite.com/docs/apis/rest-api/annotations#list-annotations-for-a-build)
+* [Artifacts](https://buildkite.com/docs/apis/rest-api/artifacts#list-artifacts-for-a-build)
+* [Builds](https://buildkite.com/docs/apis/rest-api/builds#get-a-build) (but without `jobs`, as they are stored in separate files)
+* Jobs (as would be embedded in a [Build via the REST API](https://buildkite.com/docs/apis/rest-api/builds#get-a-build))
+
+## Configure build exports
+
```
Comment (mbelton-buildkite) [inline pages/pipelines/build_exports.md:36]: Let's add a sentence here to avoid having the two headings up against each other. Perhaps:

```suggestion

To configure build exports for your organization, you'll need to prepare an S3 bucket before enabling exports in the Buildkite dashboard.

```

File: pages/pipelines/build_export.md
Comment (mbelton-buildkite) [inline pages/pipelines/build_export.md:1]: Oops, I missed that there was a URL change. It's a fairly new page, so it's probably okay but are there any links coming in here from the product? If there's uncertainty, it might be best to just keep the existing URL/filename.

**Review by mbelton-buildkite — COMMENTED** [review]
A few more small tweaks, but it's looking good 😊

File: pages/pipelines/build_export.md
Comment (123sarahj123) [reply pages/pipelines/build_export.md:1]: I don't believe so. The official feature name is build exports, so we thought to change it early when we realised the url was called build-export

Comment (123sarahj123) [reply pages/pipelines/build_export.md:1]: Confirmed with @kalopilato that build-export URL was being used, so making some changes to correct that :D

Comment (kalopilato) [reply pages/pipelines/build_export.md:1]: Yep, PR to address this [here](https://github.com/buildkite/buildkite/pull/12105), thanks for double checking folks!

**Review by mbelton-buildkite — APPROVED** [review]
Looks good to me! Great work 🙌🏻
